src/libs/blueprint/conduit_blueprint_mesh.cpp

```diff
@@ -1985,6 +1985,94 @@ mesh::number_of_domains(const conduit::Node &mesh)
     }
 }
 
+//-------------------------------------------------------------------------
+void
+mesh::state(const conduit::Node &mesh, conduit::Node &state)
+{
+    state.reset();
+    if (! is_multi_domain(mesh))
+    {
+        if (mesh.has_child("state"))
+        {
+            state.set(mesh["state"]);
+        }
+    }
+    else
+    {
+        auto doms_itr = mesh.children();
+        while (doms_itr.has_next())
+        {
+            const Node &dom = doms_itr.next();
+            if (dom.has_child("state"))
+            {
+                state.set(dom["state"]);
+                break;
+            }
+        }
+    }
+}
+
+//-------------------------------------------------------------------------
+index_t
+mesh::cycle(const conduit::Node &mesh)
+{
+    Node state;
+    ::conduit::blueprint::mesh::state(mesh, state);
+    if (state.has_child("cycle"))
+    {
+        return state["cycle"].to_index_t();
+    }
+    return std::numeric_limits<int>::max();
+}
+
+//-------------------------------------------------------------------------
+index_t
+mesh::cycle(const conduit::Node &mesh, 
+            const std::string &path,
+            bool just_asking)
+{
+    Node state;
+    ::conduit::blueprint::mesh::state(mesh, state);
+    if (state.has_child("cycle"))
+    {
+        return state["cycle"].to_index_t();
+    }
+    CONDUIT_INFO("is_cycle_present: no 'state/cycle' present."
+                 " Defaulting to counter");
+    static std::map<std::string, int> counters;
+    index_t cycle = counters[path];
+    // if we're just using this function to ask what the cycle could be, 
+    // then we don't want to mess with the static counter. If we are
+    // in the process of writing files, then we can make up a new cycle.
+    if (! just_asking)
+    {
+        counters[path] ++;
+    }
+    return cycle;
+}
```

Comment:
It is confusing to have two methods for getting the cycle, but one of them has special logic in it to handle the case where we generate a cycle number for data with no cycle. This generation is dependent on the path we are using to save data.

Reply:
I think we should handle the non cycle case outside of the cycle function.
It can happen in the gen filenames function.

Reply:
We can do it outside, but it can't just happen in the generate filenames function. The blueprint and silo writers don't use the generate filenames function.

Reply:
The issue with doing it outside is that now we need to bring back MPI communication to make sure all tasks agree on the cycle.

Reply:
Can we assume that `opts_suffix` will be the same across all tasks? Before we go about changing it depending what the cycle is. At the beginning of the function, after we've parsed the options, can we assume `opts_suffix` will be the same across all tasks?